examples/lasso/Lasso.hs

```diff
@@ -0,0 +1,189 @@
+{-# LANGUAGE DataKinds           #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeFamilies        #-}
+
+module Main where
+
+import           Data.Function                         ((&))
+import           Data.Monoid                           ((<>))
+import           Data.Singletons
+import           GHC.TypeLits
+import           Lens.Micro
+
+import           Torch.Core.Tensor.Dim                 hiding (N)
+import           Torch.Core.Tensor.Static.Double
+import           Torch.Core.Tensor.Static.DoubleMath
+import           Torch.Core.Tensor.Static.DoubleRandom
+
+-- type N = 2000 -- sample size
+-- type M = 200
+type N = 20 -- sample size
+type M = 2
+
+seedVal :: Int
+seedVal = 3141592653579
+
+genData :: TDS '[M, 1] -> IO (TDS '[N, M], TDS '[N, 1])
+genData w = do
+  gen <- newRNG
+  manualSeed gen seedVal
+  noise        :: TDS '[N, 1] <- tds_normal gen 0.0 0.10
+  predictorVal :: TDS '[N] <- tds_normal gen 0.0 1.0
+  let x :: TDS '[N, M] =
+        predictorVal
+        & tds_cat (tds_init 1.0)
+        & tds_resize
+      y = (x !*! w) ^+^ noise
+          & tds_resize :: TDS '[N, 1]
+  pure (x, y)
+
+loss :: (TDS '[N, M], TDS '[N, 1]) -> TDS '[M, 1] -> Double
+loss (x, y) w =
+  ( tds_sumAll . tds_square $ y ^-^ (x !*! w) ) + l1 w
+
+l1 :: TDS '[M, 1] -> Double
+l1 = tds_sumAll . tds_abs
+
+l2 :: TDS '[M, 1] -> Double
+l2 = tds_sumAll . tds_square
+
+prox_l1 :: TDS '[M, 1] -> Double -> TDS '[M, 1]
+prox_l1 w l        = tds_sign w ^*^ max_plus (w ^- l)
+  where max_plus t = tds_cmax t (tds_new :: TDS '[M, 1])
+
+prox_l1_single :: Double -> Double -> Double
+prox_l1_single w_i l = signum w_i * max (w_i - l) 0.0
+
+tds_getDim2d :: TDS '[N, M] -> Int -> Double
+tds_getDim2d _ i
+  | i == 0 = ((realToFrac $ natVal (Proxy :: Proxy N)) :: Double)
+  | i == 1 = ((realToFrac $ natVal (Proxy :: Proxy M)) :: Double)
+  | otherwise = error "Only works on 2d tensors; `i` must be either 0 or 1"
+
+coordinate_descent ::
```

Comment:
@AlejandroCatalina could you share a reference where this and the next CD methods are described?

Reply:
The coordinate descent algorithm for Lasso is proposed by Hastie and Tibshirani and widely covered in their book https://web.stanford.edu/~hastie/StatLearnSparsity_files/SLS_corrected_1.4.16.pdf in section 2.4